tests/contract/test_add_header.py

```diff
@@ -0,0 +1,188 @@
+        1,
+        shard_count + 1,
+        CHUNK_ROOT_1_0,
+        private_key=NotaryAccount(0).private_key
+    )
+    mine(web3, 1)
+    # Check that collation record of shard 0 has not been updated and transaction consume all gas
+    assert smc_handler.records_updated_period(0) == 0
+    assert smc_handler.get_collation_chunk_root(1, 0) == BLANK_CHUNK_ROOT
+    assert web3.eth.getTransactionReceipt(tx_hash)['gasUsed'] == default_gas
+
+    # Second attempt to add collation record with illegal shard_id specified
+    tx_hash = smc_handler.add_header(
+        1,
+        -1,
+        CHUNK_ROOT_1_0,
+        private_key=NotaryAccount(0).private_key
+    )
+    mine(web3, 1)
+    # Check that collation record of shard 0 has not been updated and transaction consume all gas
+    assert smc_handler.records_updated_period(0) == 0
+    assert smc_handler.get_collation_chunk_root(1, 0) == BLANK_CHUNK_ROOT
+    assert web3.eth.getTransactionReceipt(tx_hash)['gasUsed'] == default_gas
+
+    # Add correct collation record
+    smc_handler.add_header(1, 0, CHUNK_ROOT_1_0, private_key=NotaryAccount(0).private_key)
+    mine(web3, 1)
+    # Check that collation record of shard 0 has been updated
+    assert smc_handler.records_updated_period(0) == 1
+    assert smc_handler.get_collation_chunk_root(1, 0) == CHUNK_ROOT_1_0
+
+
+def test_double_add_header(smc_handler):  # noqa: F811
+    web3 = smc_handler.web3
+    default_gas = smc_handler.config['DEFAULT_GAS']
+
+    # Register notary 0~2 and fast forward to next period
+    batch_register(smc_handler, 0, 2)
+    fast_forward(smc_handler, 1)
+    current_period = web3.eth.blockNumber // smc_handler.config['PERIOD_LENGTH']
+    assert current_period == 1
+
+    CHUNK_ROOT_1_0 = b'\x10' * 32
+    smc_handler.add_header(1, 0, CHUNK_ROOT_1_0, private_key=NotaryAccount(0).private_key)
+    mine(web3, 1)
+    # Check that collation record of shard 0 has been updated
+    assert smc_handler.records_updated_period(0) == 1
+    assert smc_handler.get_collation_chunk_root(1, 0) == CHUNK_ROOT_1_0
+
+    # Attempt to add collation record again with same collation record
+    tx_hash = smc_handler.add_header(
+        1,
+        0,
+        CHUNK_ROOT_1_0,
+        private_key=NotaryAccount(0).private_key
+    )
+    mine(web3, 1)
+    # Check that transaction consume all gas
+    assert web3.eth.getTransactionReceipt(tx_hash)['gasUsed'] == default_gas
+
+    # Attempt to add collation record again with different chunk root
+    tx_hash = smc_handler.add_header(
+        1,
+        0,
+        b'\x56' * 32,
+        private_key=NotaryAccount(0).private_key
+    )
+    mine(web3, 1)
+    # Check that collation record of shard 0 remains the same and transaction consume all gas
+    assert smc_handler.records_updated_period(0) == 1
+    assert smc_handler.get_collation_chunk_root(1, 0) == CHUNK_ROOT_1_0
+    assert web3.eth.getTransactionReceipt(tx_hash)['gasUsed'] == default_gas
```

Comment:
Sorry, second thoughts: `HeaderAdded` log that returns `True/False` may be helpful for the client side to determine whether the `add_header` tx is included.

Reply:
Log will be emitted only if transaction succeed, i.e. passing all the checks. If client catches a `HeaderAdded` log, he can be sure that header is added.

Reply:
Currently I name it `AddHeader` to conform with other log naming pattern, e.g, `RegisterNotary`, `DeregisterNotary`.
We can change them to `HeaderAdded`, `NotaryRegistered` in the future if we like.